container/rules/debian_pkg_tar.bzl

```diff
@@ -0,0 +1,188 @@
+        image_with_keys_output_tarball = ctx.actions.declare_file(image_with_keys + ".tar")
+        image_with_keys_output_layer = ctx.actions.declare_file(image_with_keys + "-layer.tar")
+
+        _apt_key.implementation(
+            ctx,
+            name = image_with_keys,
+            image_tar = ctx.files.base[0],
+            keys = keys,
+            output_executable = image_with_keys_output_executable,
+            output_tarball = image_with_keys_output_tarball,
+            output_layer = image_with_keys_output_layer,
+        )
+        download_base = image_with_keys_output_tarball
+
+    # Declare intermediate output file generated by download_pkgs rule.
+    output_executable = ctx.actions.declare_file(ctx.attr.name + "-output_executable.sh")
+    download_pkgs_output_tar = download_pkgs_output_tar or ctx.attr.name + ".tar"
+    download_pkgs_output_script = download_pkgs_output_script or ctx.attr.name + ".sh"
+    output_tar = ctx.actions.declare_file(download_pkgs_output_tar)
+    output_script = ctx.actions.declare_file(download_pkgs_output_script)
+
+    # download_pkgs rule consumes 'packages' and 'additional_repos'.
+    _download_deb_pkgs.implementation(
+        ctx,
+        image_tar = download_base,
+        packages = packages,
+        additional_repos = additional_repos,
+        output_executable = output_executable,
+        output_tar = output_tar,
+        output_script = output_script,
+    )
+
+    return struct(
+        providers = [
+            InstallableTarInfo(
+                installables_tar = output_tar,
+            ),
+        ],
+    )
+
+def _aggregate_debian_pkgs_impl(ctx):
+    """Implementation for the aggregate_debian_pkgs rule.
+
+    aggregate_debian_pkgs rule produces a tarball with all debian packages declared
+    in the language_tool_layer(s) this rule depends on.
+
+    Args:
+      ctx: ctx as the same as for container_image + list of language_tool_layer(s)
+      https://github.com/bazelbuild/rules_docker#container_image
+    """
+
+    # If the rule is used directly, aggregate packages, additional_repos, keys
+    # from each language_tool_layer
+
+    packages = []
+    additional_repos = []
+    keys = []
+
+    for layer in ctx.attr.language_layers:
+        packages.extend(layer.packages)
+        additional_repos.extend(layer.additional_repos)
+        keys.extend(layer.keys)
+
+    packages = depset(packages).to_list()
+    additional_repos = depset(additional_repos).to_list()
+    keys = depset(keys).to_list()
+
+    return _generate_deb_tar(
+        ctx,
+        packages = packages,
+        additional_repos = additional_repos,
+        keys = keys,
+    )
+
+# Export _generate_deb_tar function for other bazel rules use.
+generate = struct(
+    attrs = debian_pkgs_attrs,
+    outputs = _download_deb_pkgs.outputs,
+    implementation = _generate_deb_tar,
+)
+
+aggregate_debian_pkgs_ = rule(
+    attrs = debian_pkgs_attrs + {
```

Comment:
have a different _attr list for aggregate_debian_pkgs_, they have nothing in common (one receives layers, the other packages and keys)

Reply:
Sure. 
One thing I notice is that if we only have ```base``` and ```language_layers``` as attributes, the  ```_download_deb_pkgs.implementation``` will cause errors, since it want to access ```additional_repos``` and ```_image_id_extractor``` if their overrides are None. I will also list those two attributes here, and validation function only allows ```name```, ```base``` and ```language_layers``` attributes.

Reply:
sgtm